libcodechecker/libhandlers/store.py

```diff
@@ -248,34 +197,77 @@ def res_handler(results):
 
 
-def process_compile_db(compile_db):
-    """
-    Create a dict where the source file name (absolute path) is the key and
-    the correspoding compilation command is the value.
-    """
-    compile_db = os.path.abspath(compile_db)
-    comp_cmds = {}
-
-    if not os.path.exists(compile_db):
-        LOG.warning("Compilation command file '" + compile_db +
-                    "' is missing.")
-        LOG.warning("Update mode will not work, without compilation "
-                    "command it is not known which file to update.")
-        return comp_cmds
-
-    with open(compile_db, 'r') as ccdb:
-        comp_cmds = json.load(ccdb)
-
-    processed = {}
-    for cc in comp_cmds:
-        file_path = os.path.join(cc['directory'], cc['file'])
+def assemble_zip(inputs, zip_file, client):
+    temp_dir = tempfile.mkdtemp()
+
+    report_dir = os.path.join(temp_dir, 'reports')
+    os.makedirs(report_dir)
+
+    source_dir = os.path.join(temp_dir, 'root')
+    os.makedirs(source_dir)
+
+    file_dict = {}
+
+    def collect_file_hashes_from_plist(plist_file):
         try:
-            # Newer compile command json format.
-            processed[file_path] = str(' '.join(cc['arguments']))
-        except Exception:
-            processed[file_path] = cc['command']
+            files, _ = plist_parser.parse_plist(plist_file)
+
+            for f in files:
+                with open(f) as content:
+                    hasher = sha256()
+                    hasher.update(content.read())
+                    content_hash = hasher.hexdigest()
+                    file_dict[content_hash] = f
+        except Exception as ex:
+            LOG.error('Parsing the plist failed: ' + str(ex))
 
-    return processed
+    try:
+        for input_path in inputs:
+            input_path = os.path.abspath(input_path)
+
+            if os.path.isfile(input_path):
+                if not input_path.endswith(".plist"):
+                    continue
+                shutil.copy(input_path, report_dir)
+                collect_file_hashes_from_plist(input_path)
+            elif os.path.isdir(input_path):
+                _, _, files = next(os.walk(input_path), ([], [], []))
+                for f in files:
+                    plist_file = os.path.join(input_path, f)
+                    shutil.copy(plist_file, report_dir)
+                    if f.endswith(".plist"):
+                        collect_file_hashes_from_plist(plist_file)
+
+        necessary_hashes = client.necessaryFileContents(file_dict.keys())
+        for h, f in file_dict.items():
+            if h in necessary_hashes:
+                target_file = os.path.join(source_dir, f[1:])
+
+                try:
+                    os.makedirs(os.path.dirname(target_file))
+                except os.error:
+                    # Directory already exists.
+                    pass
+
+                shutil.copyfile(f, target_file)
+
+        zipf = zipfile.ZipFile(zip_file, 'w')
```

Comment:
The .zip file is compressed later in one piece when it's already containing everything. I'm not writing the compressed file to the disk to a `.zip.gz`, but I'm sending the data directly to the server.